cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -118,15 +126,40 @@ protected BatchSparkPipelineDriver(BatchSQLEngineAdapter sqlEngineAdapter) {
   protected SparkCollection<RecordInfo<Object>> getSource(StageSpec stageSpec,
                                                           FunctionCache.Factory functionCacheFactory,
                                                           StageStatisticsCollector collector) {
-    PluginFunctionContext pluginFunctionContext = new PluginFunctionContext(stageSpec, sec, collector);
-    FlatMapFunction<Tuple2<Object, Object>, RecordInfo<Object>> sourceFunction =
-      new BatchSourceFunction(pluginFunctionContext, functionCacheFactory.newCache());
+    // Initialize function cache factory
     this.functionCacheFactory = functionCacheFactory;
-    return new RDDCollection<>(sec, functionCacheFactory, jsc,
-                               new SQLContext(jsc), datasetContext, sinkFactory, sourceFactory
-      .createRDD(sec, jsc, stageSpec.getName(), Object.class, Object.class)
-      .flatMap(sourceFunction)
-    );
+
+    String sourceStageName = stageSpec.getName();
+
+    // Supplier for Spark backed implementation
+    Supplier<SparkCollection<RecordInfo<Object>>> sparkCollectionSupplier =
+      getSparkCollectionSupplier(stageSpec, collector);
+
+    // If the SQL Engine is initialized, and the stage is a SQL Engine compatible Input stage, return a Future
+    // Collection which will try to execute the SQL Input operation before falling back into the Spark implementation.
+    if (sqlEngineAdapter != null && sourceFactory.getSQLEngineInput(sourceStageName) != null) {
+      SQLEngineInput sourceSQLEngineInput = sourceFactory.getSQLEngineInput(sourceStageName);
+
+      // Create read task to read using the SQL engine input
+      Supplier<SQLBackedCollection<RecordInfo<Object>>> sourceReadTask = getSQLSourceReadTask(sourceStageName,
```

Comment:
I'd better make a call here if we are reading direct instead of delaying it. Are there any reason we need to delay the decision?

Reply:
Because the read execution might succeed or not. I do this to ensure we do not "block" the pipeline everytime we need to wait for a read task to complete.

Reply:
Well, I don't think that failed execution warrants this complexity. As soon as we decide that the execution is supported, it should go with pushdown. Otherwise we may get unstable plans suddenly switching between the ways to do things.

Reply:
Overall I think we should decide the execution plan before actually running the execution. Otherwise we'll have huge problems deciding "what" to pushdown after each node (if we allow each node to fall back from pushing down during the execution)

Reply:
Reads are not guaranteed to succeed. For example, we might not have permissions to read from the source dataset, in which case the read operation would fail.

>Otherwise we'll have huge problems deciding "what" to pushdown after each node (if we allow each node to fall back from pushing down during the execution)

Not really. Let's say we have a BQ Source and a Group-By Aggregation which follows.

1. If pushdown is enabled and we were able to direct-read from source, then the Group By runs in BQ.
2. If pushdown is enabled and the direct-read from source failed, we read records using the standard workflow and run in Spark.

Reply:
Exactly. What if the failure is intermittent? This would switch between two modes with unpredicted consequences. It's much better to error out any access problems and let user fix it than silently switch to fallback and make it very hard to find out what's wrong

Reply:
With out strategy getting a source out of pushdown may mean job would "never" finish, especially on a small cluster not suited for data processing. And it would be very hard to find out why.

Reply:
The problem with requiring all sources to be compatible for Direct reads is that it could be a problem for larger pipelines with many sources, particularily with BigQuery Pushdown:

1. All sources would need to be in the same **Location**
2. All source projects, datasets and tables need to be accessible from the service account configured for Pushdown execution.

>And it would be very hard to find out why.

We can always add logging to show the users why something is failing (wrong location, permissions issue, etc).